Index: ntsilk/ffmpeg/libavcodec/codec_id.h
===================================================================
--- ntsilk.orig/ffmpeg/libavcodec/codec_id.h
+++ ntsilk/ffmpeg/libavcodec/codec_id.h
@@ -545,6 +545,8 @@ enum AVCodecID {
     AV_CODEC_ID_QOA,
     AV_CODEC_ID_LC3,
 
+    AV_CODEC_ID_NTSILK_S16LE = 0x15ff0,
+
     /* subtitle codecs */
     AV_CODEC_ID_FIRST_SUBTITLE = 0x17000,          ///< A dummy ID pointing at the start of subtitle codecs.
     AV_CODEC_ID_DVD_SUBTITLE = 0x17000,
