Index: ntsilk/ffmpeg/libavcodec/allcodecs.c
===================================================================
--- ntsilk.orig/ffmpeg/libavcodec/allcodecs.c
+++ ntsilk/ffmpeg/libavcodec/allcodecs.c
@@ -505,6 +505,7 @@ extern const FFCodec ff_mpc8_decoder;
 extern const FFCodec ff_msnsiren_decoder;
 extern const FFCodec ff_nellymoser_encoder;
 extern const FFCodec ff_nellymoser_decoder;
+extern const FFCodec ff_ntsilk_skp_s16le_encoder;
 extern const FFCodec ff_on2avc_decoder;
 extern const FFCodec ff_opus_encoder;
 extern const FFCodec ff_opus_decoder;
