Index: ntsilk/ffmpeg/libavformat/allformats.c
===================================================================
--- ntsilk.orig/ffmpeg/libavformat/allformats.c
+++ ntsilk/ffmpeg/libavformat/allformats.c
@@ -324,6 +324,10 @@ extern const FFInputFormat  ff_nc_demuxe
 extern const FFInputFormat  ff_nistsphere_demuxer;
 extern const FFInputFormat  ff_nsp_demuxer;
 extern const FFInputFormat  ff_nsv_demuxer;
+extern const FFInputFormat  ff_ntsilk_skp_s16le_demuxer;
+extern const FFOutputFormat ff_ntsilk_skp_s16le_muxer;
+extern const FFInputFormat  ff_ntsilk_tct_s16le_demuxer;
+extern const FFOutputFormat ff_ntsilk_tct_s16le_muxer;
 extern const FFOutputFormat ff_null_muxer;
 extern const FFInputFormat  ff_nut_demuxer;
 extern const FFOutputFormat ff_nut_muxer;
